From 9fe242243d43ed8f4252b8bdd55300a53a581426 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Sat, 19 Jul 2014 04:29:01 -0400
Subject: [acpi] pci: Use global PCI rescan-remove locking in PCI root hotplug

Message-id: <20140719042901.19438.13937.stgit@gir.stowe>
Patchwork-id: 86022
O-Subject: [RHEL7.1 PATCH 60/81] ACPI / PCI: Use global PCI rescan-remove locking in PCI root hotplug
Bugzilla: 1110895
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110895
Upstream Status: 7a3bb55ebd1cb04f08dd68cf60af24b434f8cc2a

commit 7a3bb55ebd1cb04f08dd68cf60af24b434f8cc2a
Author: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
Date:   Fri Jan 10 15:23:14 2014 +0100

    ACPI / PCI: Use global PCI rescan-remove locking in PCI root hotplug

    Multiple race conditions are possible between the addition and removal of
    PCI devices during ACPI PCI host bridge hotplug and the generic PCI bus
    rescan and device removal that can be triggered via sysfs.

    To avoid those race conditions make the ACPI PCI host bridge addition and
    removal code use global PCI rescan-remove locking.

    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/acpi/pci_root.c b/drivers/acpi/pci_root.c
index 20360e4..5b01bd6 100644
--- a/drivers/acpi/pci_root.c
+++ b/drivers/acpi/pci_root.c
@@ -599,7 +599,9 @@ static int acpi_pci_root_add(struct acpi_device *device,
   pci_assign_unassigned_root_bus_resources(root->bus);
  }
 
+ pci_lock_rescan_remove();
  pci_bus_add_devices(root->bus);
+ pci_unlock_rescan_remove();
  return 1;
 
 end:
@@ -611,6 +613,8 @@ static void acpi_pci_root_remove(struct acpi_device *device)
 {
  struct acpi_pci_root *root = acpi_driver_data(device);
 
+ pci_lock_rescan_remove();
+
  pci_stop_root_bus(root->bus);
 
  device_set_run_wake(root->bus->bridge, false);
@@ -618,6 +622,8 @@ static void acpi_pci_root_remove(struct acpi_device *device)
 
  pci_remove_root_bus(root->bus);
 
+ pci_unlock_rescan_remove();
+
  kfree(root);
 }
 
-- 
1.7.1